internal/utils/temporary_utils.go

```diff
@@ -0,0 +1,118 @@
+// TO BE REMOVED AFTER INTEGRATING ADDON INSTANCE CLIENT SDK
+
+package utils
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"reflect"
+	"time"
+
+	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func SetAddonInstanceCondition(ctx context.Context, cacheBackedKubeClient client.Client, condition metav1.Condition, addonName string) error {
+	addonInstance, err := getAddonInstanceByAddon(ctx, cacheBackedKubeClient, addonName)
+	if err != nil {
+		return fmt.Errorf("failed to fetch AddonInstance by the addon '%s': %w", addonName, err)
+	}
+	if err := upsertAddonInstanceCondition(ctx, cacheBackedKubeClient, &addonInstance, condition); err != nil {
+		return fmt.Errorf("failed to update the conditions of the AddonInstance resource in the namespace %s: %w", addonInstance.Namespace, err)
+	}
+	return nil
+}
+
+type ReconcilerWithHeartbeat interface {
+	reconcile.Reconciler
+	GetLatestHeartbeat() metav1.Condition
+	SetLatestHeartbeat(metav1.Condition)
+}
+
+func SetupHeartbeatReporter(r ReconcilerWithHeartbeat, mgr manager.Manager, addonName string, handleAddonInstanceConfigurationChanges func(addonsv1alpha1.AddonInstanceSpec)) error {
+	defaultHealthyHeartbeatConditionToBeginWith := metav1.Condition{
+		Type:    "addons.managed.openshift.io/Healthy",
+		Status:  "True",
+		Reason:  "AddonHealthy",
+		Message: fmt.Sprintf("Addon '%s' is perfectly healthy", addonName),
+	}
+	// initialized with a healthy heartbeat condition corresponding to the addon
+	r.SetLatestHeartbeat(defaultHealthyHeartbeatConditionToBeginWith)
```

Comment:
I don't quite get this.
An Addon needs to decide on it's own whether it's ready or not, ideally by observing the clusters state and reporting based on that information.
Defaulting the state to success make the first reported state completely useless, as it's not reporting the actual state of the addon.

So Addons will get installed, be Healthy once and than fail directly afterwards....

Defaulting to an Unkown or Unsuccessful state in the beginning makes more sense to me.

Reply:
Good point!

My perspective was that for any addon in production, it would have passed the osde2e tests in the past because if not, then, it wouldn't have been gated to production in the first place. And, succeeding osde2e tests are indicative of the fact that at least the addon successfully gets installed and bootstrapped on an OSD cluster.

But yeah this poses a risk around the edge case that if an addon with not-so-good heartbeat reporting setup, it might fail, say, just after 30 seconds of getting installed, and say, and still the heartbeat-reporter-central-loop might continue to report the default healthy heartbeat.

Thanks for raising this, I'll make the updates.

That's why it felt like a safe assumption to make that for the first time, it's ok to report a successful heartbeat. But yeah, if the addon starts falling apart, say, 30 seconds after its installation, it will (should) still have the logic to report a failure heartbeat, and if that's not the case, then it will atleast end up reporting HeartbeatTimeout.